docs/source/model/models-from-code.rst

```diff
@@ -0,0 +1,511 @@
+
+            with mlflow.start_run():
+                model_info = mlflow.pyfunc.log_model(
+                    python_model=model_path,  # Define the model as the path to the script that was just saved
+                    artifact_path="arithemtic_model",
+                    input_example=[42.0, 24.0]
+                )
+
+
+        Looking at this stored model within the MLflow UI, we can see that the script in the first cell was recorded as an artifact to the run. 
+        
+        .. figure:: ../_static/images/models/basic_model_from_code_ui.png
+            :alt: The MLflow UI showing the stored model code as a serialized python script
+            :width: 80%
+            :align: center
+
+        When we load this model via ``mlflow.pyfunc.load_model()``, this script will be executed and an instance of ``BasicModel`` will be constructed, exposing the ``predict`` 
+        method as our entry point for inference, just as with the alternative legacy mode of logging a custom model.
+
+        .. code-block:: python
+            
+            my_model = mlflow.pyfunc.load_model(model_info.model_uri)
+            my_model.predict([2.2,3.1,4.7]) 
+
+            # or, with a Pandas DataFrame input
+            my_model.predict(pd.DataFrame([5.0,6.0,7.0]))
+    
+    .. tab:: Models with Code Paths dependencies
+
+        In this example, we will explore a more complex scenario that demonstrates how to work with multiple Python scripts and leverage the ``code_paths`` 
+        feature in MLflow for model management. Specifically, we will define a ``Calculator`` class in one script, which performs basic arithmetic 
+        operations, and then use this class within an ``ArithmeticModel`` custom ``PythonModel`` that we will define in a separate script. 
+        This model will be logged with MLflow, allowing us to perform predictions using the stored model.
+
+        This tutorial will show you how to:
+
+        - Create multiple Python files from within a Jupyter notebook.
+        - Log a custom model with MLflow that relies on external code defined in another file.
+        - Use the ``code_paths`` feature to include additional scripts when logging the model, ensuring that all dependencies are available when the model is loaded for inference.
+
+        In the first step, we define a ``Calculator`` class in a file named ``calculator.py``. This class includes a basic arithmetic operation, such as 
+        adding two numbers, and also keeps a history of operations performed. The purpose of this class is to encapsulate the logic that will be used 
+        later in the MLflow model.
+
+        The following code block writes the ``Calculator`` class definition to ``calculator.py``:
+        
+        .. code-block:: python
+
+            %%writefile "./calculator.py"
+
+            from typing import List, TypeVar
+
+
+            T = TypeVar('T', int, float, complex)
+
+            class Calculator():
```

Comment:
Can we simplify the example? For example:

```
# add.py

def add(x, y):
    return x + y
```

```
# main_model

from add import add

class MyModel:

    # use `add`
```

Reply:
Good call. It is a bit superfluous to add in the additional logic. Updating!